scipy/fftpack/czt.py

```diff
@@ -0,0 +1,457 @@
+        The ratio between points in each step.  This must be precise or the
+        accumulated error will degrade the tail of the output sequence.
+    a : complex, optional
+        The starting point in the complex plane.  The default is 1+0j.
+
+    Returns
+    -------
+    f : CZT
+        callable object ``f(x, axis=-1)`` for computing the chirp z-transform
+        on `x`
+
+    See Also
+    --------
+    ZoomFFT : for a friendly interface to partial FFT calculations
+
+    Notes
+    -----
+    If `w` does not lie on the unit circle, then the transform will be
+    around a spiral with exponentially-increasing radius.  Regardless,
+    angle will increase linearly.
+
+    For transforms that do lie on the unit circle, accuracy is better when
+    using `ZoomFFT`, since any numerical error in `w` is
+    accumulated for long data lengths, drifting away from the unit circle.
+
+    The chirp z-transform can be faster than an equivalent FFT with
+    zero padding.  Try it with your own array sizes to see.
+
+    However, the chirp z-transform is considerably less precise than the
+    equivalent zero-padded FFT, with differences on the order of 1e-7 from the
+    direct transform rather than 1e-15 as seen with zero-padding.
+
+    As this CZT is implemented using the Bluestein algorithm, it can compute
+    large prime-length Fourier transforms in O(N log N) time, rather than the
+    O(N**2) time required by the direct DFT calculation used in `fft`.
+
+    References
+    ----------
+    .. [1] Leo I. Bluestein, "A linear filtering approach to the computation
+           of the discrete Fourier transform," Northeast Electronics Research
+           and Engineering Meeting Record 10, 218-219 (1968).
+    .. [2] Rabiner, Schafer, and Rader, "The chirp z-transform algorithm and
+           its application," Bell Syst. Tech. J. 48, 1249-1292 (1969).
+
+    Examples
+    --------
+    Compute multiple prime-length FFTs:
+
+    >>> a = np.random.rand(7)
+    >>> b = np.random.rand(7)
+    >>> c = np.random.rand(7)
+    >>> czt_7 = CZT(n=7)
+    >>> A = czt_7(a)
+    >>> B = czt_7(b)
+    >>> C = czt_7(c)
+
+    Display the points at which the FFT is calculated:
+
+    >>> czt_7.points()
+    array([ 1.00000000+0.j        ,  0.62348980+0.78183148j,
+           -0.22252093+0.97492791j, -0.90096887+0.43388374j,
+           -0.90096887-0.43388374j, -0.22252093-0.97492791j,
+            0.62348980-0.78183148j])
+    >>> import matplotlib.pyplot as plt
+    >>> plt.plot(np.real(czt_7.points()), np.imag(czt_7.points()), 'o')
+    >>> plt.show()
+
+    """
+    def __init__(self, n, m=None, w=None, a=1+0j):
+        m = _validate_sizes(n, m)
+
+        k = arange(max(m, n), dtype=np.min_scalar_type(-max(m, n)**2))
+
+        if w is None:
+            # Nothing specified, default to FFT-like
+            w = cmath.exp(-2j*pi/m)
+            wk2 = np.exp(-(1j * pi * k**2) / m)
+        else:
+            # w specified
+            wk2 = w**(k**2/2.)
```

Comment:
Similar situation to the one a few lines above. Only here we unfortunately don't know how to do the "manual range reduction".

If the user had a way to specify `w` not as a complex number, but rather as a tuple of "absolute value" and "angle given as a fraction of 2pi", we could do the same thing here. I admit that this kind of interface is awkward, but it could do wonders for accuracy.

Reply:
Alternatively one could try to convert arg(w)/(2pi) into a fraction using an algorithm like https://www.ics.uci.edu/~eppstein/numth/frap.c

Reply:
I think [`float.as_integer_ratio()`](https://docs.python.org/3/library/stdtypes.html#float.as_integer_ratio) can be used to do that.
I agree that doing such a range reduction is necessary for good accuracy. In a CZT implementation I did once, I converted the user-specified grid spacing to a fraction and used that for range reduction. I also specified a `snap=True` kwarg that allowed the algorithm to choose a slightly different grid spacing (within a few ulp) if that had a fraction representation with a small denominator (otherwise the range reduction has no advantage).

Reply:
> I think float.as_integer_ratio() can be used to do that.

Doesn't look like it, unfortunately:
```
In [5]: (0.1).as_integer_ratio()
Out[5]: (3602879701896397, 36028797018963968)
```
It's important to be able to specify a maximum denominator.
Snapping to a fraction with a sufficiently small (TBD) denominator should be acceptable if the error is less than 1ulp. If that's not possible, then I'd assume that the user did not intend to specify an exact fraction of 2pi, and one should probably fall back to the current implementation.

Reply:
One way to limit the denominator is implemented in the standard library in the `fractions` module via the [`Fractions.limit_denominator`](https://docs.python.org/3/library/fractions.html#fractions.Fraction.limit_denominator). Checking if there is a nice decimal fraction within 1 ulp seems like a very sensible idea to me.

Reply:
I think this thread was resolved with [the discussion](https://github.com/scipy/scipy/pull/4607#issuecomment-713035772) about specifying `w` explicitly in pieces rather than trying to figure out an angle from a float.  

If no one else does, I will try to create a new PR for adding support for `w=(rad, nom, denom)` or `w=(rad, Fraction(nom, denom))`.